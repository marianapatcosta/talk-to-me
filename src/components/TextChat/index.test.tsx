import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import { TextChat } from './index'

const input = 'This is a message'
const placeholder = 'text here'
const defaultOutput = "I didn't understand"
const onTypingMock = jest.fn()
const onUnmatchedOutputMock = jest.fn()

describe('TextChat', () => {
  const defaultProps = {
    conversation: [
      {
        inputKeywords: [],
        inputs: [input],
        output: 'Successful test!',
      },
    ],
    defaultOutput,
    textChatPlaceholder: placeholder,
    onTyping: onTypingMock,
    onUnmatchedOutput: onUnmatchedOutputMock,
  }
  
  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<TextChat {...defaultProps} />)

    expect(firstChild).toMatchSnapshot()
  })

  it('should display the corresponding output, if the provided input is in `conversation`', async () => {
    render(<TextChat {...defaultProps} />)
    const textarea = screen.getByPlaceholderText(
      placeholder
    ) as HTMLTextAreaElement
    fireEvent.click(textarea, { target: { value: input } })
    expect(textarea.value).toBe(input)
    fireEvent.click(screen.getByRole('button'))

    expect(onTypingMock).toHaveBeenCalled()
    expect(onUnmatchedOutputMock).not.toHaveBeenCalled()

    const output = await screen.findByText('Successful test!', undefined, {
      timeout: 5000,
    })
    expect(output).toBeInTheDocument()
  })

  it('should display the defaultOutput, if the provided input is not in `conversation`', async () => {
    const randomInput = 'This is random'
    render(<TextChat {...defaultProps} />)
    const textarea = screen.getByPlaceholderText(
      placeholder
    ) as HTMLTextAreaElement
    fireEvent.click(textarea, { target: { value: randomInput } })
    expect(textarea.value).toBe(randomInput)
    fireEvent.click(screen.getByRole('button'))

    const output = await screen.findByText(defaultOutput, undefined, {
      timeout: 5000,
    })
    expect(onTypingMock).toHaveBeenCalled()
    expect(onUnmatchedOutputMock).toHaveBeenCalled()
    expect(output).toBeInTheDocument()
  })
})
