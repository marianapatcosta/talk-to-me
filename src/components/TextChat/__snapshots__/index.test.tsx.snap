// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextChat renders without errors and matches snapshot 1`] = `
<div
  class="undefined undefined"
>
  <div
    class="undefined  "
  >
    <form>
      <textarea
        placeholder="text here"
        rows="2"
      />
      <button
        aria-label="click to submit your message"
        class="undefined "
        type="submit"
      >
        <svg
          data-testid="send-svg"
          role="img"
          viewBox="0 0 32 32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            send
          </title>
          <path
            d="M4.667 26.307v-7.983L21.81 16.02 4.667 13.716V5.733l24 10.285z"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
